User Page (frontend)

alonweissfeld opened the pull request:
This PR is the continuation of #38 and it resolves #7.

It introduces the `user.html` template and it's css to display a simple user page. Some more nicer layout might be added in the future (with #35), but currently, it displays the user name, past and future streams. 

If the user is having a live stream at the moment (of the requested time of the page), it will display a red-dot next to his name and show the relevant live link.

For more implementation details on the view that serves this template, see #38.

Thanks.

Commit:
Added the user.html template and user.css for displaying the user info in addition to his previous and future streams.

Commit:
Added the relevant html and tempalte conditions to show an appropriate message when a user exist but has no streams.
Note that the main_app namespace was added to the main urls.py file of the project.

tarelmadar reviewed (approved):
I've reviewed the files, looks fine to me.

lmilbaum reviewed (changes requested):
I don't see any test included in the PR

rosomri commented on sellotape/sellotape_dj/main_app/templates/sellotape/user.html:
```diff
@@ -0,0 +1,64 @@
+                    <li>
+                        <h4>{{ stream.title }}</h4>
+                        <p>
+                            Aired on {{ stream.airs_on|date }} <a href="{{ stream.link }}">here</a>. <br/>
```
Don't we prefer to embed the streams on our site?
By identifying the platform (e.g. YouTube and Twitch) and using their embedding format.

rosomri commented on sellotape/sellotape_dj/main_app/templates/sellotape/user.html:
```diff
@@ -0,0 +1,64 @@
+                    <li>
+                        <h4>{{ stream.title }}</h4>
+                        <p>
+                            Airs on {{ stream.airs_on|date }} <a href="{{ stream.link }}">here</a>.
```
I think we should embed the streams on our website,
the way to do it is by identifying the platform (e.g. YouTube and Twitch) or by adding a 'platform' field to 'Streams DB',
and then using platform's embedding format.

alonweissfeld replied:
That could be cool, but in my opinion we should consider it later as an enhancement feature and first have a simpler platform. The Stream Model can always be modified :)

alonweissfeld commented:
@lioramilbaum The tests are part of #38. View them [here](https://github.com/beyond-io/sellotape/pull/38/files#diff-45acd7c2db27ba3f92007e5ade877ec6).

I've split the task into two PRs - one is the backend work and tests for the appropriate view and template handling and this is the 2nd one regarding the frontend. It's hard for me to think about further tests for the HTML/CSS here, since the templating engine is responsible for displaying it right, but if you have any ideas let me know :)

Also, once this is merged, more tests could be added that link between the expected html and the context data passed to the template (For example, at #38, [this line is commented out](https://github.com/beyond-io/sellotape/pull/38/files#diff-45acd7c2db27ba3f92007e5ade877ec6R41) since it is split from it's HTML.

lioramilbaum commented:
> @lioramilbaum The tests are part of #38. View them [here](https://github.com/beyond-io/sellotape/pull/38/files#diff-45acd7c2db27ba3f92007e5ade877ec6).

In general, each PR should consume the functionality and the test(s) which validates the functionality. It is required for maintainers who are not familiar with the technology. The tests which are executed by the PR provides confidence in the change.

> 
> I've split the task into two PRs - one is the backend work and tests for the appropriate view and template handling and this is the 2nd one regarding the frontend. It's hard for me to think about further tests for the HTML/CSS here, since the templating engine is responsible for displaying it right, but if you have any ideas let me know :)

frontend tests are usually implemented with tools such as selenium. 

> 
> Also, once this is merged, more tests could be added that link between the expected html and the context data passed to the template (For example, at #38, [this line is commented out](https://github.com/beyond-io/sellotape/pull/38/files#diff-45acd7c2db27ba3f92007e5ade877ec6R41) since it is split from it's HTML.

You are referring to the integration tests or as we call them e2e (end to end) tests. Those can be provided by their own dedicated PR.

Commit:
Added tests to the user template.

Mock a context object and render the template, then test against the expected html.

alonweissfeld commented:
@lioramilbaum 
> You are referring to the integration tests or as we call them e2e (end to end) tests. Those can be provided by their own dedicated PR. 

I don't think these are e2e tests as they test the view mostly and what is expected from the returned html by using Model mocks. However, I've added some testing _explicitly to the template functionality_ (as related to this PR) without involving the view.

For example, if I pass a `live_stream` variable, I'm expecting some html tag and text to show it.
Other than that, selenium seems like an overkill to be honest since there is no further UI functionality needed to be tested here (html elements aren't affecting each other, no js involved, just text we expect to see from the html).

Please take a look, thank you

lioramilbaum commented:
> @lioramilbaum
> 
> > You are referring to the integration tests or as we call them e2e (end to end) tests. Those can be provided by their own dedicated PR.
> 
> I don't think these are e2e tests as they test the view mostly and what is expected from the returned html by using Model mocks. However, I've added some testing _explicitly to the template functionality_ (as related to this PR) without involving the view.
> 
> For example, if I pass a `live_stream` variable, I'm expecting some html tag and text to show it.
> Other than that, selenium seems like an overkill to be honest since there is no further UI functionality needed to be tested here (html elements aren't affecting each other, no js involved, just text we expect to see from the html).
> 
> Please take a look, thank you

@alonweissfeld I admit that it is has been a long working day for me. Too tired to be drill down in the details. I liked your clear and fierce explanation. It is enough for me.
